File: app/api/stats.py
"""Aggregate processing/usage statistics for the dashboard."""

from __future__ import annotations

import re
from collections import defaultdict

from fastapi import APIRouter, Depends
from sqlmodel import Session, func, select

from app.db import get_session
from app.models import Item, ItemStatus, StageName, StageRun, Transcript
from app.schemas import PlatformStat, StatsRead

router = APIRouter(prefix="/api/stats", tags=["stats"])

# Each CJK character counts as one "word"; each run of latin letters/digits as
# one word. A pragmatic word count for mixed Chinese/English transcripts.
_WORD_RE = re.compile(r"[\u4e00-\u9fff]|[0-9A-Za-z]+")


@router.get("", response_model=StatsRead)
def get_stats(session: Session = Depends(get_session)) -> StatsRead:
    total_items = session.exec(select(func.count()).select_from(Item)).one()

    by_status: dict[str, int] = defaultdict(int)
    for status, count in session.exec(
        select(Item.status, func.count()).group_by(Item.status)
    ).all():
        by_status[status.value] = count

    by_platform: dict[str, int] = defaultdict(int)
    for platform, count in session.exec(
        select(Item.platform, func.count()).group_by(Item.platform)
    ).all():
        by_platform[platform.value] = count

    avg_stage: dict[str, float] = {}
    total_stage: dict[str, float] = {}
    cost_by_stage: dict[str, float] = {}
    for stage, avg_ms, sum_ms, sum_cost in session.exec(
        select(
            StageRun.stage,
            func.avg(StageRun.duration_ms),
            func.sum(StageRun.duration_ms),
            func.sum(StageRun.cost_usd),
        ).group_by(StageRun.stage)
    ).all():
        avg_stage[stage.value] = float(avg_ms or 0)
        total_stage[stage.value] = float(sum_ms or 0)
        cost_by_stage[stage.value] = float(sum_cost or 0.0)

    def _sum(column, *stages: StageName) -> int:
        stmt = select(func.coalesce(func.sum(column), 0))
        if stages:
            stmt = stmt.where(StageRun.stage.in_([s for s in stages]))
        return int(session.exec(stmt).one() or 0)

    # --- Source media length + transcript volume ---
    total_duration_s = float(
        session.exec(select(func.coalesce(func.sum(Item.duration_s), 0))).one() or 0
    )
    transcript_words = 0
    transcript_chars = 0
    for text in session.exec(select(Transcript.text)).all():
        if text:
            transcript_chars += len(text)
            transcript_words += len(_WORD_RE.findall(text))

    # --- Per-platform detail (items + duration from Item, cost/tokens via join) ---
    platforms: dict[str, dict] = {}
    for platform, count, dur in session.exec(
        select(
            Item.platform,
            func.count(),
            func.coalesce(func.sum(Item.duration_s), 0),
        ).group_by(Item.platform)
    ).all():
        platforms[platform.value] = {
            "items": int(count or 0),
            "duration_s": float(dur or 0),
            "done": 0,
            "tokens": 0,
            "cost_usd": 0.0,
        }
    for platform, count in session.exec(
        select(Item.platform, func.count())
        .where(Item.status == ItemStatus.done)
        .group_by(Item.platform)
    ).all():
        if platform.value in platforms:
            platforms[platform.value]["done"] = int(count or 0)
    for platform, tokens, cost in session.exec(
        select(
            Item.platform,
            func.coalesce(func.sum(StageRun.total_tokens), 0),
            func.coalesce(func.sum(StageRun.cost_usd), 0.0),
        )
        .select_from(StageRun)
        .join(Item, StageRun.item_id == Item.id)
        .group_by(Item.platform)
    ).all():
        if platform.value in platforms:
            platforms[platform.value]["tokens"] = int(tokens or 0)
            platforms[platform.value]["cost_usd"] = float(cost or 0.0)
    by_platform_detail = [
        PlatformStat(platform=name, **data)
        for name, data in sorted(
            platforms.items(), key=lambda kv: kv[1]["items"], reverse=True
        )
    ]

    openrouter_requests = _sum(StageRun.request_count, StageName.transcribe)
    openrouter_tokens = _sum(StageRun.total_tokens, StageName.transcribe)
    gemini_tokens = _sum(StageRun.total_tokens, StageName.summarize, StageName.gemini_audio)
    http_429_total = _sum(StageRun.http_429_count)
    prompt_tokens = _sum(StageRun.prompt_tokens)
    completion_tokens = _sum(StageRun.completion_tokens)
    total_tokens = _sum(StageRun.total_tokens)
    total_cost = float(session.exec(
        select(func.coalesce(func.sum(StageRun.cost_usd), 0.0))
    ).one() or 0.0)

    return StatsRead(
        total_items=int(total_items or 0),
        items_by_status=dict(by_status),
        items_by_platform=dict(by_platform),
        by_platform=by_platform_detail,
        avg_stage_ms=avg_stage,
        total_stage_ms=total_stage,
        cost_by_stage=cost_by_stage,
        total_duration_s=total_duration_s,
        transcript_words=transcript_words,
        transcript_chars=transcript_chars,
        prompt_tokens=prompt_tokens,
        completion_tokens=completion_tokens,
        total_tokens=total_tokens,
        openrouter_requests=openrouter_requests,
        openrouter_tokens=openrouter_tokens,
        gemini_tokens=gemini_tokens,
        total_cost_usd=total_cost,
        http_429_total=http_429_total,
    )

File: cf/worker/src/routes/stats.ts
import { Hono } from "hono";
import type { AppContext } from "../auth";
import { requireAuth } from "../auth";
import { all, first } from "../db";

export const statsRoutes = new Hono<AppContext>();
statsRoutes.use("*", requireAuth);

// Aggregate stats over the global catalog + pipeline runs.
statsRoutes.get("/", async (c) => {
  const refresh = c.req.query("refresh") === "true";
  const total = (await first<{ n: number }>(c.env.DB.prepare("SELECT COUNT(*) AS n FROM item")))?.n ?? 0;

  const byStatus = await all<{ status: string; n: number }>(
    c.env.DB.prepare("SELECT status, COUNT(*) AS n FROM item GROUP BY status"),
  );
  const byPlatform = await all<{ platform: string; n: number }>(
    c.env.DB.prepare("SELECT platform, COUNT(*) AS n FROM item GROUP BY platform"),
  );
  const stages = await all<{ stage: string; n: number; avg_ms: number; total_ms: number; cost: number }>(
    c.env.DB.prepare(
      `SELECT stage, COUNT(*) AS n, AVG(duration_ms) AS avg_ms,
              SUM(duration_ms) AS total_ms, SUM(cost_usd) AS cost
         FROM stage_run GROUP BY stage`,
    ),
  );
  const tokens = await first<{ prompt: number; completion: number; total: number; cost: number; r429: number; reqs: number }>(
    c.env.DB.prepare(
      `SELECT SUM(prompt_tokens) AS prompt, SUM(completion_tokens) AS completion,
              SUM(total_tokens) AS total, SUM(cost_usd) AS cost,
              SUM(http_429_count) AS r429, SUM(request_count) AS reqs
         FROM stage_run`,
    ),
  );
  const dur = (await first<{ d: number }>(c.env.DB.prepare("SELECT SUM(duration_s) AS d FROM item")))?.d ?? 0;

  const items_by_status: Record<string, number> = {};
  for (const r of byStatus) items_by_status[r.status] = r.n;
  const items_by_platform: Record<string, number> = {};
  for (const r of byPlatform) items_by_platform[r.platform] = r.n;
  const avg_stage_ms: Record<string, number> = {};
  const total_stage_ms: Record<string, number> = {};
  const cost_by_stage: Record<string, number> = {};
  for (const s of stages) {
    avg_stage_ms[s.stage] = s.avg_ms ?? 0;
    total_stage_ms[s.stage] = s.total_ms ?? 0;
    cost_by_stage[s.stage] = s.cost ?? 0;
  }

  if (refresh) c.header("cache-control", "no-store");
  return c.json({
    total_items: total,
    items_by_status,
    items_by_platform,
    by_platform: byPlatform.map((p) => ({ platform: p.platform, items: p.n, done: 0, duration_s: 0, tokens: 0, cost_usd: 0 })),
    avg_stage_ms,
    total_stage_ms,
    cost_by_stage,
    total_duration_s: dur,
    transcript_words: 0,
    transcript_chars: 0,
    prompt_tokens: tokens?.prompt ?? 0,
    completion_tokens: tokens?.completion ?? 0,
    total_tokens: tokens?.total ?? 0,
    openrouter_requests: tokens?.reqs ?? 0,
    openrouter_tokens: 0,
    gemini_tokens: 0,
    total_cost_usd: tokens?.cost ?? 0,
    http_429_total: tokens?.r429 ?? 0,
  });
});
